Faster generation using AWQ + Fused modules

younesbelkada opened the pull request:
# What does this PR do?

Introduces a new feature - fused module generation using `autoawq` library. Users need to specify modules that they want to fuse inside `fusing_mapping`. 

The API is as follows:

```python
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM, AwqConfig, TextStreamer

model_id = "TheBloke/Mistral-7B-OpenOrca-AWQ"
torch_device = "cuda" if torch.cuda.is_available() else "cpu"

quantization_config = AwqConfig(
    bits=4,
    do_fuse=True,
    fuse_max_seq_len=512,
)

model = AutoModelForCausalLM.from_pretrained(model_id, quantization_config=quantization_config).to(0)
tokenizer = AutoTokenizer.from_pretrained(model_id)

streamer = TextStreamer(tokenizer, skip_prompt=True, skip_special_tokens=True)

prompt_template = """\
<|im_start|>system
You are MistralOrca, a large language model trained by Alignment Lab AI. Write out your reasoning step-by-step to be sure you get the right answers!<|im_end|>
<|im_start|>user
{prompt}<|im_end|>
<|im_start|>assistant"""

prompt = "You're standing on the surface of the Earth. "\
        "You walk one mile south, one mile west and one mile north. "\
        "You end up exactly where you started. Where are you?"

tokenizer.pad_token = tokenizer.eos_token

inputs = tokenizer([prompt_template.format(prompt=prompt), prompt_template.format(prompt=prompt), prompt_template.format(prompt=prompt)], return_tensors="pt", padding=True).to(0)

outputs = model.generate(**inputs, max_new_tokens=512)
print(tokenizer.batch_decode(outputs, skip_special_tokens=True))
```

## Before this PR:

![before-fusing](https://github.com/huggingface/transformers/assets/49240599/6086e7ad-d00a-427b-a2ce-30ac53b2efe4)

## After this PR:

![fast-awq-generate](https://github.com/huggingface/transformers/assets/49240599/f0c1fd61-34a0-4c26-b04e-54c6df9554ba)

## TODOs:

- [x] Block save_pretrained if one creates a fused model
- [x] Fuse MLP
- [x] wait until dynamic batch size is resolved on auto-awq side : https://github.com/casper-hansen/AutoAWQ/pull/181
- [x] Force users to use `autoawq>=0.1.7`
- [x] Tests
- [x] Benchmarks
- [x] Docs

cc @amyeroberts @casper-hansen @SunMarc

younesbelkada committed:
v1 fusing modules

younesbelkada committed:
add fused mlp support

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_27411). All of your documentation changes will be reflected on that endpoint.

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into awq-fused-modules

younesbelkada committed:
up

younesbelkada committed:
fix CI

younesbelkada committed:
block save_pretrained

younesbelkada committed:
fixup

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into awq-fused-modules

younesbelkada committed:
small fix

younesbelkada committed:
add new condition

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into awq-fused-modules

younesbelkada committed:
Merge branch 'awq-fused-modules' of https://github.com/younesbelkada/transformers into awq-fused-modules

younesbelkada committed:
add v1 docs

younesbelkada committed:
add some comments

younesbelkada commented:
Before moving forward with tests and advanced docs, I would love to have an early feedback of the API that is described in the PR description. cc @amyeroberts , whenever you have time, i would appreciate your feedback on this PR 🙏 Thanks!

younesbelkada committed:
Merge branch 'main' into awq-fused-modules

amyeroberts commented on src/transformers/integrations/awq.py:
```diff
@@ -102,3 +120,145 @@ def replace_with_awq_linear(
+    return current_fused_mapping
+
+
+def fuse_awq_modules(model, quantization_config):
```
This is a pretty big function - I'd split it up so that there's private functions for each layer-replacement which is then called within the big for-loop

younesbelkada replied:
Done in [cde53ef](https://github.com/huggingface/transformers/pull/27411/commits/cde53efe376ef47ba49fe27c92b722161fd52d88)

amyeroberts replied:
Beautiful 🤩

amyeroberts commented on src/transformers/integrations/awq.py:
```diff
@@ -102,3 +120,145 @@ def replace_with_awq_linear(
+                ).to(old_module.weight.device)
+                del old_module
+        # Replace MLP layers
+        if hasattr(module, fusing_mapping["mlp"][0]):
```
Can we guarantee `fusing_mapping` has an `"mlp"` key? and that it has at least one value?

younesbelkada replied:
I propose to do a more extensive check in b187c07 and I changed the logic there a bit to make sure these methods will do nothing in case we put an empty array in these fields

amyeroberts commented on src/transformers/integrations/awq.py:
```diff
@@ -102,3 +120,145 @@ def replace_with_awq_linear(
+                del old_module
+        # Replace MLP layers
+        if hasattr(module, fusing_mapping["mlp"][0]):
+            gate_proj = getattr(module, fusing_mapping["mlp"][0])
+            up_proj = getattr(module, fusing_mapping["mlp"][1])
+            down_proj = getattr(module, fusing_mapping["mlp"][2])
```
gate_, up_ down_ is highly specific to just a few models. Instead we can generalise this to take a list of linear layers so that any MLP can be passed

amyeroberts commented on src/transformers/integrations/awq.py:
```diff
@@ -102,3 +120,145 @@ def replace_with_awq_linear(
+            down_proj = getattr(module, fusing_mapping["mlp"][2])
+
+            previous_device = gate_proj.qweight.device
+            activation_fn = ACT2FN[model.config.hidden_act]
```
Same comment here re config param names. `hidden_act` is common but it's not always the name used. Could we pass this in in the fuse config too?

younesbelkada replied:
Hmm, do you think my suggestion here: https://github.com/huggingface/transformers/pull/27411#discussion_r1400819721 would be applied in this case?

amyeroberts commented on src/transformers/integrations/awq.py:
```diff
@@ -102,3 +120,145 @@ def replace_with_awq_linear(
+            if isinstance(q_proj, WQLinear_GEMV):
+                target_cls = WQLinear_GEMV
+                cat_dim = 0
+            else:
+                target_cls = WQLinear_GEMM
+                cat_dim = 1
```
```suggestion
            elif isinstance(q_proj, WQLinear_GEMM):
                target_cls = WQLinear_GEMM
                cat_dim = 1
            else: 
                raise ValueError("Unsupported q_proj type: {type(q_proj)}")
```

amyeroberts reviewed (commented):
Really nice piece of work! 🔥 

Main comment is about structure of the input arguments. It might just be my current understanding from the PR so please do correct me if I'm wrong. 

At the moment, it seems there's two types of arguments which configure the behaviour: the fusing_mapping and max_seq_len. Does `max_seq_len` have its own config value because we expect this to be the value users modify a lot? 

For `AwqConfig` do we foresee other `fuse` arguments to be added to configure this behaviour? If so, we might want to bundle them all together into a single `fuse_config` which the `AwqConfig` owns.

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into awq-fused-modules

younesbelkada committed:
style

younesbelkada committed:
fix nit

younesbelkada committed:
adapt from suggestion

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2782,7 +2778,10 @@ def from_pretrained(
+        if (
+            quantization_method_from_config == QuantizationMethod.GPTQ
+            or quantization_method_from_config == QuantizationMethod.AWQ
+        ) and quantization_method_from_args is not None:
```
Fixed in [14c820d](https://github.com/huggingface/transformers/pull/27411/commits/14c820d3b6c044160762af9dac8bbd281c4f2d62)

younesbelkada commented on src/transformers/integrations/awq.py:
```diff
@@ -102,3 +120,145 @@ def replace_with_awq_linear(
+
+    Args:
+        model (`~PreTrainedModel`):
+            The model to fuse - note this model should have been converted into AWQ format beforehand.
```
Fixed in [03d8dff](https://github.com/huggingface/transformers/pull/27411/commits/03d8dff668a826852b8772ef50f68e68f471aaa5)

younesbelkada committed:
add check

younesbelkada commented on src/transformers/integrations/awq.py:
```diff
@@ -102,3 +120,145 @@ def replace_with_awq_linear(
+        num_key_value_heads = getattr(model.config, "num_key_value_heads", num_attention_heads)
+
+        # Fill `current_fused_mapping` with the expected values
+        current_fused_mapping["hidden_size"] = hidden_size
+        current_fused_mapping["num_attention_heads"] = num_attention_heads
+        current_fused_mapping["num_key_value_heads"] = num_key_value_heads
+        current_fused_mapping["max_seq_len"] = quantization_config.fuse_max_seq_len
```
I propose to do it in [0a08551](https://github.com/huggingface/transformers/pull/27411/commits/0a08551c0c82c63b823055f81b8b3de1c031e691) lmk what do you think!

younesbelkada commented on src/transformers/integrations/awq.py:
```diff
@@ -102,3 +120,145 @@ def replace_with_awq_linear(
+        current_fused_mapping["hidden_size"] = hidden_size
+        current_fused_mapping["num_attention_heads"] = num_attention_heads
+        current_fused_mapping["num_key_value_heads"] = num_key_value_heads
+        current_fused_mapping["max_seq_len"] = quantization_config.fuse_max_seq_len
```
I think `quantization_config` cannot be `None` here no? Except if users want to call the public method directly, but in that case `quantization_config` is a required arg so those users should know what they are doing.
If you want I can also add a check to make sure the passed config is not None;

amyeroberts replied:
Misreading the if condition above. My bad - sorry!

younesbelkada commented on src/transformers/utils/quantization_config.py:
```diff
@@ -543,6 +543,12 @@ class AwqConfig(QuantizationConfigMixin):
         backend (`AwqBackendPackingMethod`, *optional*, defaults to `AwqBackendPackingMethod.AUTOAWQ`):
             The quantization backend. Some models might be quantized using `llm-awq` backend. This is useful for users
             that quantize their own models using `llm-awq` library.
+        fuse_modules (`bool`, *optional*, defaults to `False`):
+            Whether to fuse attention and mlp layers together for faster inference
+        fuse_max_seq_len (`int`, *optional*):
+            The Maximum sequence length to generate when using fusing.
+        fusing_mapping (`dict`, *optional*, default to `None`):
+            Overwrite the natively supported fusing scheme with the one specified by the users.
```
Agreed, `do_fuse` and `modules_to_fuse` sounds great!

younesbelkada committed:
change arg names

younesbelkada committed:
change variables name

younesbelkada commented on src/transformers/utils/quantization_config.py:
```diff
@@ -587,3 +599,14 @@ def post_init(self):
+        # TODO: awq version check for fused modules.
+
+    @property
+    def is_using_fused_modules(self):
```
Yes! Changed it

younesbelkada committed:
Update src/transformers/integrations/awq.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

younesbelkada committed:
style

younesbelkada committed:
split up into 3 different private methods

younesbelkada committed:
more conditions

younesbelkada committed:
more checks

younesbelkada committed:
add fused tests for custom models

younesbelkada committed:
fix

younesbelkada commented on src/transformers/utils/quantization_config.py:
```diff
@@ -587,3 +599,14 @@ def post_init(self):
+
+    @property
+    def is_using_fused_modules(self):
+        return self.fuse_modules or (self.fusing_mapping is not None and len(self.fusing_mapping) > 0)
```
Makes sense! Modified it according to your suggestion

younesbelkada committed:
fix tests

younesbelkada committed:
final update docs

younesbelkada commented:
Thank you very much @amyeroberts for your extensive review! I addressed most of your comments and left some open questions - let me know what do you think! 🙏

younesbelkada committed:
final fixes

younesbelkada commented on src/transformers/utils/quantization_config.py:
```diff
@@ -587,3 +604,42 @@ def post_init(self):
+        if self.do_fuse:
+            awq_version_supports_fusing = False
+            MIN_AWQ_VERSION = "0.1.7"
+            if is_auto_awq_available():
+                # For some reason `version.parse(importlib.metadata.version("awq"))` always returns
+                # `<Version('0.1.0')>` which makes that logic unusable. Therefore we need to import
+                # awq and get `awq.__version__`
+                import awq
+
+                awq_version = awq.__version__
+                awq_version_supports_fusing = version.parse(awq_version) >= version.parse(MIN_AWQ_VERSION)
```
@casper-hansen do you have an idea why `importlib.metadata.version("awq")` always gives me `0.1.0` ?

casper-hansen replied:
I have no idea, `0.1.0` is nowhere to be found in AutoAWQ.

https://github.com/casper-hansen/AutoAWQ/blob/main/awq/__init__.py#L1

Is it picking up llm-awq?

https://github.com/mit-han-lab/llm-awq/blob/main/pyproject.toml

younesbelkada replied:
Indeed this works ! Thanks very much @casper-hansen !

younesbelkada committed:
fix importlib metadata

amyeroberts commented on src/transformers/utils/quantization_config.py:
```diff
@@ -587,3 +604,38 @@ def post_init(self):
+            ]
+            if not all(key in self.modules_to_fuse for key in required_keys):
+                raise ValueError(
+                    f"Required fiels are missing in the fusing mapping, required fields are {required_keys}"
```
```suggestion
                    f"Required fields are missing in the fusing mapping, required fields are {required_keys}"
```

amyeroberts commented on src/transformers/utils/quantization_config.py:
```diff
@@ -587,3 +604,38 @@ def post_init(self):
+                    "You current version of `autoawq` does not support module fusing, please upgrade `autoawq` package."
+                )
+
+        if self.do_fuse and self.modules_to_fuse is not None:
+            required_keys = [
+                "hidden_size",
+                "num_attention_heads",
+                "num_key_value_heads",
+                "mlp",
+                "attention",
+                "layernorm",
+                "use_alibi",
+            ]
```
Based on these keys it would be cool to have a tool which automatically generates a config for a model assuming you wanted to fuse all modules.

amyeroberts reviewed (approved):
Thanks for iterating! 

Great code, great tests, great docs 💯

SunMarc commented on src/transformers/utils/quantization_config.py:
```diff
@@ -543,6 +543,12 @@ class AwqConfig(QuantizationConfigMixin):
             that quantize their own models using `llm-awq` library.
+        do_fuse (`bool`, *optional*, defaults to `False`):
+            Whether to fuse attention and mlp layers together for faster inference
+        fuse_max_seq_len (`int`, *optional*):
```
```suggestion
        fuse_max_seq_len (`int`, *optional*, defaults to `None`):
```

SunMarc commented on src/transformers/utils/quantization_config.py:
```diff
@@ -587,3 +604,38 @@ def post_init(self):
+                )
+
+            if not awq_version_supports_fusing:
+                raise ValueError(
+                    "You current version of `autoawq` does not support module fusing, please upgrade `autoawq` package."
+                )
```
```suggestion
                raise ValueError(
                    f"You current version of `autoawq` does not support module fusing, please upgrade `autoawq` package to at least {MIN_AWQ_VERSION}."
                )
```

SunMarc commented on src/transformers/utils/quantization_config.py:
```diff
@@ -587,3 +604,38 @@ def post_init(self):
+                "layernorm",
+                "use_alibi",
+            ]
+            if not all(key in self.modules_to_fuse for key in required_keys):
```
Would be great also to check that the `attention` key have 0 or exactly 3 values.

SunMarc commented on src/transformers/integrations/awq.py:
```diff
@@ -102,3 +121,214 @@ def replace_with_awq_linear(
+    else:
+        raise ValueError("Fusing is only supported for the AutoAWQ backend")
+
+    for name, module in model.named_modules():
+        # Replace layer norms
+        _fuse_awq_layernorm(modules_to_fuse["layernorm"], module, FasterTransformerRMSNorm)
+
+        # Replace MLP layers
+        _fuse_awq_mlp(model, name, modules_to_fuse["mlp"], module, QuantFusedMLP)
+
+        # Replace attention layers
+        _fuse_awq_attention_layers(model, module, modules_to_fuse, name, QuantAttentionFused)
```
Right now, there is no warning if any module was replaced or not. Maybe add a warning somewhere if the fusing was not successful (because the user didn't put the right name).

SunMarc commented on src/transformers/utils/quantization_config.py:
```diff
@@ -587,3 +604,38 @@ def post_init(self):
+                    f"Required fiels are missing in the fusing mapping, required fields are {required_keys}"
+                )
+
+        # TODO: awq version check for fused modules.
```
Is it about MIN_AWQ_VERSION = "0.1.7" ? if so, we can remove that since it is done. Otherwise, we should do it if it is important. 
```suggestion
```

SunMarc commented on src/transformers/modeling_utils.py:
```diff
@@ -2809,13 +2801,10 @@ def from_pretrained(
             quantization_method_from_config = config.quantization_config.get(
                 "quant_method", QuantizationMethod.BITS_AND_BYTES
             )
-        if quantization_method_from_config is not None and quantization_method_from_args is not None:
-            if quantization_method_from_config != quantization_method_from_args:
-                raise ValueError(
-                    f"The model is already quantized with {quantization_method_from_config}. "
-                    f"You can't quantize it again with {quantization_method_from_args}"
-                )
```
Please keep [this](https://github.com/huggingface/transformers/pull/27500). This is in case someone tries to combine different quantization scheme together (e.g bitsandbytes + gptq)

SunMarc commented on src/transformers/modeling_utils.py:
```diff
@@ -2704,17 +2707,6 @@ def from_pretrained(
-                quantization_config, "quant_method", QuantizationMethod.BITS_AND_BYTES
-            )
-
-            if quantization_method_from_args == QuantizationMethod.AWQ:
-                raise ValueError(
-                    "You cannot pass an `AwqConfig` when loading a model as you can only use AWQ models"
-                    " for inference. To quantize transformers models with AWQ algorithm, please refer to our"
-                    " quantization docs: https://huggingface.co/docs/transformers/main_classes/quantization "
-                )
```
We can keep this but just check is the model is not already quantized by checking `quantization_method_from_config == QuantizationMethod.AWQ`

younesbelkada replied:
Makes sense, I added that logic in [0697687](https://github.com/huggingface/transformers/pull/27411/commits/06976877955b57078ffec1acec7aa25d17d1714c) with an additional test in [12aff7c](https://github.com/huggingface/transformers/pull/27411/commits/12aff7c3a52ad0d693e07b03cc98dde30a0c999d)

SunMarc commented on src/transformers/modeling_utils.py:
```diff
@@ -2809,13 +2801,10 @@ def from_pretrained(
+            quantization_method_from_config in (QuantizationMethod.GPTQ, QuantizationMethod.AWQ)
+            and quantization_method_from_args is not None
+        ):
             loading_attr_dict = quantization_config.get_loading_attributes()
             for attr, val in loading_attr_dict.items():
```
I see that `get_loading_attributes` is not defined in `AWQConfig` . If it is not defined, we won't be able to overwrite the quantization config stored in config, hence performing the fusing. It is strange that it worked.

younesbelkada replied:
Hmm indeed it was not defined before, I just added it now

SunMarc commented on tests/quantization/autoawq/test_awq.py:
```diff
@@ -225,3 +209,117 @@ def test_quantized_model_multi_gpu(self):
+        with self.assertRaises(ValueError), tempfile.TemporaryDirectory() as tmpdirname:
+            model.save_pretrained(tmpdirname)
+
+    def test_generation_fused(self):
```
let's also check if we indeed have the fused modules. We can't see that when testing the generation.

SunMarc commented on tests/quantization/autoawq/test_awq.py:
```diff
@@ -225,3 +209,117 @@ def test_quantized_model_multi_gpu(self):
+
+        self.assertEqual(tokenizer.decode(outputs[0], skip_special_tokens=True), self.EXPECTED_GENERATION)
+
+    def test_generation_fused_batched(self):
```
check fused modules here as well

younesbelkada replied:
Fixed in [e65687b](https://github.com/huggingface/transformers/pull/27411/commits/e65687b7488dd950c3fc4ab6736f5bb421967875)

SunMarc commented on tests/quantization/autoawq/test_awq.py:
```diff
@@ -225,3 +209,117 @@ def test_quantized_model_multi_gpu(self):
+        self.assertEqual(tokenizer.decode(outputs[0], skip_special_tokens=True), self.EXPECTED_GENERATION)
+
+    @require_torch_multi_gpu
+    def test_generation_custom_model(self):
```
check fused modules here as well

younesbelkada replied:
Fixed in [e65687b](https://github.com/huggingface/transformers/pull/27411/commits/e65687b7488dd950c3fc4ab6736f5bb421967875)

SunMarc reviewed (commented):
Thanks for making AWQ faster through fused modules 🔥. The design looks great and could be easily extended to other quantization scheme in the future. I left a few comments.

SunMarc commented on src/transformers/modeling_utils.py:
```diff
@@ -3560,6 +3549,14 @@ def from_pretrained(
                 pass
 
+        if (
+            quantization_config is not None
+            and quantization_config.quant_method == QuantizationMethod.AWQ
+            and quantization_config.do_fuse
+        ):
+            model = fuse_awq_modules(model, quantization_config)
+            model._awq_is_fused = True
+
```
I see that you are using the quantization_config that was passed in the `from_pretrained`. However, if the fused parameters was stored in the `config.quantization_config`, we won't be able to perform the fusing. This is why I decided to overwrite the `config.quantization_config` using` get_loading_attributes()`. What could be done here is use config.quantization_config instead. LMK what you think.

SunMarc replied:
Yes, we should always fuse if the user passes do_fuse or any args that are in `get_loading_attributes()`. Thanks to `get_loading_attributes()`, we overwrite `config.quantization_config` to have information about `do_fuse` or other args. 

What I wanted is to use only one config for quantization. Right now, we have the configuration that was passed and the config that was stored. The merging of the two configurations through `get_loading_attributes()` was motivated by the fact that some args should not be modified such as the `backend` and the `version`.

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into awq-fused-modules

younesbelkada committed:
Merge branch 'awq-fused-modules' of https://github.com/younesbelkada/transformers into awq-fused-modules

younesbelkada committed:
Update src/transformers/utils/quantization_config.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

younesbelkada committed:
change it to `do_fuse`

younesbelkada committed:
nit

younesbelkada commented on src/transformers/integrations/awq.py:
```diff
@@ -102,3 +121,214 @@ def replace_with_awq_linear(
+            use_alibi=modules_to_fuse["use_alibi"],
+        )
+
+        fused_attention_layer.is_hf_transformers = True
```
Yes, it is used in auto-awq, check out https://github.com/casper-hansen/AutoAWQ/pull/195

younesbelkada committed:
Update src/transformers/utils/quantization_config.py

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

younesbelkada committed:
Update src/transformers/utils/quantization_config.py

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

younesbelkada committed:
Update src/transformers/utils/quantization_config.py

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

younesbelkada committed:
Merge branch 'awq-fused-modules' of https://github.com/younesbelkada/transformers into awq-fused-modules

younesbelkada committed:
few fixes

younesbelkada committed:
revert

younesbelkada committed:
fix test

younesbelkada committed:
fix copies

SunMarc reviewed (approved):
Thanks for iterating on this ! I've left a few comments about two points but they are not blocking for this PR !

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into awq-fused-modules

younesbelkada committed:
raise error if model is not quantized

younesbelkada committed:
add test

younesbelkada committed:
use quantization_config.config when fusing

younesbelkada commented on src/transformers/modeling_utils.py:
```diff
@@ -2830,21 +2827,36 @@ def from_pretrained(
+            and quantization_method_from_config is None
+        ):
+            raise ValueError(
+                "You cannot quantize a non-quantized model using transformers, please refer to the quantization documentation"
```
```suggestion
                "You cannot quantize with AWQ a non-quantized model using transformers, please refer to the quantization documentation"
```

younesbelkada committed:
Update src/transformers/modeling_utils.py
